File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第8章Hankson的趣味题.cpp
// lanqiaoOJ题号520

#include<bits/stdc++.h>
using namespace std;
int lcm(int a, int b){ return a / __gcd(a, b) * b;}
int main() {
    int n; scanf("%d",&n);
    while(n--) {
        int a0,a1,b0,b1;
        cin >>a0>>a1>>b0>>b1;
        int ans=0;
        for(int x=1;x <= sqrt(b1);x++){
            if(b1%x == 0){        //优化
                if(__gcd(x,a0)==a1 && lcm(x,b0)==b1)  ans++;
                int y = b1/x;       //另外一个因子
                if(x==y) continue;
                if(__gcd(y,a0)==a1 && lcm(y,b0)==b1)  ans++;
            }
        }
        cout << ans <<endl;
    }
    return 0;
}

File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第8章Hankson的趣味题.java
//Hankson 的趣味题  lanqiaoOJ  520 

import java.io.*;
import java.util.*;

public class Main {

    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static Scanner cin = new Scanner(System.in);
    public static int gcd(int a , int b) {
        return b > 0 ? gcd(b , a % b) : a;
    }
    public static int lcm(int a , int b){
        return a / gcd(a , b) * b;
    }
    public static void main(String[] args) {
        int T = cin.nextInt();
        while(T -- > 0){
            int a0 = cin.nextInt() , a1 = cin.nextInt() , b0 = cin.nextInt() , b1 = cin.nextInt();
            int ans = 0;
            for(int x = 1 ; x * x <= b1 ; x ++){
                if(b1 % x == 0) {
                    if(gcd(x , a0) == a1 && lcm(x , b0) == b1) ans ++ ;
                    int y = b1 / x;
                    if(x == y) continue ;
                    if(gcd(y , a0) == a1 && lcm(y , b0) == b1) ans ++;
                }
            }
            System.out.println(ans);
        }
    }
}